pkg/ui/client.go

```diff
@@ -0,0 +1,187 @@
+	type Alias struct {
+		Key          string                 `json:"key"`
+		Label        string                 `json:"label"`
+		IsDependency bool                   `json:"isDependency"`
+		DependencyOf []string               `json:"dependencyOf"`
+		Data         map[string]interface{} `json:"data"`
+	}
+	alias := &Alias{}
+	if err := json.Unmarshal(data, alias); err != nil {
+		return fmt.Errorf("error during Application.UnmarshalJSON: %v\n", err)
+	}
+
+	dependencyOf := map[string]interface{}{}
+	for _, appName := range alias.DependencyOf {
+		dependencyOf[appName] = struct{}{}
+	}
+
+	*this = Application{
+		Key:          alias.Key,
+		Label:        alias.Label,
+		IsDependency: alias.IsDependency,
+		DependencyOf: dependencyOf,
+		Data:         alias.Data,
+	}
+
+	return nil
+}
+
+// Client struct
+type Client struct {
+	ctx    *cli.Context
+	client api.Client
+}
+
+func (this *Client) Token() string {
+	conf := config.Read()
+
+	return conf.Token
+}
+
+func (this *Client) Project() *manifest.ProjectManifest {
+	project, err := manifest.FetchProject()
+	if err != nil {
+		return nil
+	}
+
+	return project
+}
+
+func (this *Client) Context() *manifest.Context {
+	context, err := manifest.FetchContext()
+	if err != nil {
+		return nil
+	}
+
+	return context
+}
+
+func (this *Client) Install(applications []Application, domains, buckets []string) error {
+	path := manifest.ContextPath()
+	context, err := manifest.ReadContext(path)
+	if err != nil {
+		context = manifest.NewContext()
+	}
+
+	this.addDomains(context, domains)
+	this.addBuckets(context, buckets)
+
+	installableApplications := algorithms.Filter(applications, func(app Application) bool {
+		return !app.IsDependency
+	})
+
+	dependencies := algorithms.Filter(applications, func(app Application) bool {
+		return app.IsDependency
+	})
+
+	for _, dep := range dependencies {
+		if err = this.doInstall(dep, context); err != nil {
+			return err
+		}
+	}
+
+	for _, app := range installableApplications {
+		if err = this.doInstall(app, context); err != nil {
+			return err
+		}
+	}
+
+	// Write to context.yaml only if there were no errors
+	err = context.Write(path)
+	return err
+}
+
+func (this *Client) doInstall(application Application, context *manifest.Context) error {
+	recipeID := application.Data["id"].(string)
+	oidc := application.Data["oidc"].(bool)
+	configuration := application.Data["context"].(map[string]interface{})
+	repoName := application.Label
+	mergedConfiguration, exists := context.Configuration[repoName]
+	if !exists {
+		mergedConfiguration = map[string]interface{}{}
+	}
+
+	recipe, err := this.client.GetRecipeByID(recipeID)
+	if err != nil {
+		return api.GetErrorResponse(err, "GetRecipeByID")
+	}
+
+	// Merge incoming configuration with existing one and update context
+	maps.Copy(mergedConfiguration, configuration)
+	context.Configuration[repoName] = mergedConfiguration
+
+	// Configure OIDC if enabled
+	if oidc {
```

Comment:
this should happen after installing the recipe i believe

Reply:
ahh, I thought it doesn't do any API mutations so I moved it there. I see it does the `UpsertOidcProvider` after all. I will change it then.